python_scripts/load_arrow.py

```diff
@@ -1,13 +1,295 @@
+        same name.
+        """
+        col_name_index = self.column_names.index(column_name)
+        self.arrow_table = self.arrow_table.remove_column(col_name_index)
+        self.arrow_table = self.arrow_table.add_column(
+            col_name_index, column_name, converted
+        )
+
+    def replace_missing_values(self):
+        """
+        Replace null values in the arrow table with the value stata needs.
+        Dates/timestamps/category require additional conversion before loading to
+        stata, so null replacements are deferred till later.
+        """
+        for column_name, column_type in self.column_types.items():
+            if column_type in ["date", "timestamp", "category"]:
+                # Skip dates/timestamps/category
+                # We replace these after conversion to stata dates in `load_data`
+                continue
+            null_value = self.MISSING_VALUES[column_type]
+            converted = self.arrow_table[column_name].fill_null(null_value)
+            col_name_index = self.column_names.index(column_name)
+            self.arrow_table = self.arrow_table.remove_column(col_name_index)
+            self.arrow_table = self.arrow_table.add_column(
+                col_name_index, column_name, converted
+            )
+
+    def get_categorical_encodings(self, varnames):
+        """
+        Retrieve a mapping of labels: values for categorical vars.
+        Returns a dict where the variable names are the keys and the value
+        is a dictionary consisting of label : value key/value pairs
+        """
+        value_labels = {}
+        for varname in varnames:
+            first_chunk = self.arrow_table[varname].chunks[0]
+            # Get the unique values for this variable and numeric mapping
+            value_map = {
+                str(value): idx for idx, value in enumerate(first_chunk.dictionary)
+            }
+            value_labels[varname] = value_map
+
+        return value_labels
+
+    def make_vars(self):
+        """
+        Function used to create the Stata variables in memory.
+        This assumes that the values in name_and_type object are valid
+        Stata variable names.
+        :param name_and_type: A dictionary structured to contain the variable names
+        as the keys and the storage type of the variable as the value.
+        """
+        # Loop over the variable name / variable type mappings and add
+        # variables with the appropriate typing based on the vartype
+        for varname, vartype in self.column_types.items():
+            if vartype == "string":
+                sfi.Data.addVarStrL(varname)
+            elif vartype == "boolean":
+                sfi.Data.addVarByte(varname)
+            elif vartype == "int":
+                sfi.Data.addVarInt(varname)
+            elif vartype == "category":
+                sfi.Data.addVarLong(varname)
+            elif vartype in ["float", "date", "timestamp"]:
```

Comment:
It does, but it also can't deal with big ints, so if there's a timestamp that becomes a very big number when we convert it to an int, I think it would need to be a float.  Timestamps can be stored as number of milliseconds since 1st Jan 1960, and I don't think stata can cope with that as a integer type.

Reply:
So that's definitely true for dates and at the moment dates are all we support in ehrQL so we should be able to store those as `int32` and everything should be fine.

But looking again at timestamps, I think Stata's docs are just internally inconsistent here:
![image](https://user-images.githubusercontent.com/19630/232725971-c2641b26-4fb0-4543-ac79-782f8bb54138.png)

The example integer they give is bigger than the biggest integer type you can store in Stata. So I think timestamps must be stored as floats, with the assumption that the loss of precision doesn't matter for the purposes for which they're used.

So I guess we have to store timestamps as floats. Or we could punt on the issue, given that we don't yet emit timestamps from ehrQL.